[1주차] 영화 목록 조회 API

zhdiddl opened the pull request:
### 제목(title)
영화 목록 조회 API 작성

### 작업 내용  
- 멀티 모듈 설계 및 아키텍처 구성
- 영화 목록 조회 기능에 필요한 엔티티, 컨트롤러, 레포지토리 구현

### 발생했던 문제와 해결 과정을 남겨 주세요. 
- 문제: 멀티 모듈 구조로 작성하면서 스프링 부트가 엔티티 및 레포지토리를 제대로 스캔하지 못하는 상황이 발생
- 해결 방법: 명시적인 설정을 가진 애너테이션을 추가해 스캔 범위 지정
```java
@EntityScan(basePackages = "com.example.moviedomain.entity") // 엔티티 스캔
@EnableJpaRepositories(basePackages = "com.example.movieadapter.db") // 레포지토리 스캔
@ComponentScan(basePackages = {"com.example"}) // 모든 모듈의 빈 스캔
@SpringBootApplication
public class MovieApiApplication {

	public static void main(String[] args) {
		SpringApplication.run(MovieApiApplication.class, args);
	}

}
```

### 이번 주차에서 고민되었던 지점이나, 어려웠던 점을 알려 주세요.
-  **아키텍처 선택**: 레이어드 아키텍처만 사용하다가 이번 기회에 클린 아키텍처와 헥사고날 아키텍처에 대해 알아보게 되었습니다. 두 아키텍처가 다소 비슷한 부분이 있다고 느껴져셔 트레이드 오프를 고려해 선택하기까지 시간이 많이 소요됐는데요. Domain이 외부 변화에 대한 의존성을 최소화 하도록 설계해서 시스템의 확장성과 유연성을 보장해야 한다는 요구 사항을 고려해, 좀 더 구조가 세분화된 헥사고날을 선택해서 구현해 보았습니다. 의존성 역전 원칙과 단일 책임 원칙에 가까운 구조로서 차후 기능을 확장했을 때도 안정적일 수 있겠다는 생각이 들었지만, 구조가 세분화된 만큼 작성해야 하는 코드 양과 개발 시간이 많이 증가한다는 느낌을 받았습니다. 패키지를 다 구성하고 나서도 개발 중에 위치나 역할을 헷갈리는 순간이 자주 있었고, 만약 팀으로 작업할 시 해당 아키텍처에 익숙하지 않다면 소통에도 문제가 있지 않을까 하는 생각이 들었습니다. 결과적으로 다 완성하고 나니, port나 adapter라는 요소를 사용해 헥사고날 아키텍처라고 할 수도 있지만, 레이어드 아키텍처처럼 각 계층이 하향식으로 의존하는 형태로 구현이 된 것 같기도 합니다.

### 리뷰 포인트
- 헥사고날 아키텍처를 적용한 현재 모듈 및 패키지 구조가 올바른지 궁금합니다.

### 기타 질문
[아키텍처 관련 질문]
- **UseCase 적용**: 헥사고날 아키텍처에 UseCase라는 요소가 있다고 하는데 활용이 잘 와닿지 않아 구현 시 적용하지 못했습니다. 추상화 역할을 하는 port와 비슷한 것 같은데 구체적으로 어떤 경우에 사용하면 좋을지, 현재 구조에서도 이걸 적용하면 더 나은 아키텍처가 될 지 궁금합니다.
- **`api` 모듈에서 `adapter` 직접 의존**: api 모듈이 application 모듈만 의존하도록 구현하려고 했으나 application의 세부 구현체 역할인 adapter 모듈에도 의존해야 하는 상황이 됐습니다. adapter 모듈에 대한 의존성을 제거하면 MovieRepository의 구현체를 찾을 수 없어서 애플리케이션 실행에 오류가 발생했는데요. adapter 모듈의 빈이 스캔되지 않은 문제인가 해서 @ComponentScan(basePackages = "com.example")을 추가해보았으나 해결되지 않았습니다. 개선하는 것이 좋다면 어떻게 개선이 가능한지 궁금합니다.
- **`adapter` 모듈에서 `application` 직접 의존**: DB 관련 구현체인 adapter 모듈이 application 모듈을 의존하지 않고 domain 모듈에만 의존하는 게 이상적일 것 같다고 생각했는데, application 모듈의 port 사용을 위해 의존성을 추가했습니다. 이러한 구성에 문제가 있을지 궁금합니다.
- **`application.yml` 파일 위치**: 애플리케이션 진입점을 api 모듈에 두고 application.yml 파일을 추가해 개발용 환경 설정 파일을 활성화를 설정하고, application.yml을 개발용과 운영용으로 각각 생성해 상세 설정 정보를 작성했는데, 해당 상세 정보에 DB 연결 정보가 포함되어 파일 위치를 DB 의존성을 가진 adapter 모듈에 배치하게 됐습니다. 환경 설정 파일을 두 개 모듈에 둔 현재 구현이 적절한지 궁금합니다.

[코드 구현 관련 질문]
- 엔티티에서 `@Column(nullable = false)`를 설정할 때 어떤 기준으로 추가 여부를 판단하는면 좋을까요? 예를 들어, 영화 엔티티를 작성하면서 저는 영화 제목 필드에 해당 설정을 생략하고 싶었는데, 이유는 생성자를 팩토리 메서드로 직접 제공하면서 값을 반드시 입력하라는 의도를 이미 표현했다고 생각했기 때문이었습니다. 그래도 각 컬럼에 이 속성을 명시적으로 붙이는 것이 더 현업에서 지향하는 방법일까요?
- 엔티티 생성자를 of 같은 메서드명을 가진 팩토리 메서드로 제공하는 것에 대한 트레이드 오프를 어떻게 생각하시는지 궁금합니다. new를 통한 객체 생성 방식은 잘 쓰지 않다는 의견, 팩토리 메서드로 객체 생성을 제어하는 것이 좋다는 의견, 팩토리 메서드는 depth가 깊어서 좋지 않다는 의견 등등 다양한 의견을 접했는데, 어떤 기준으로 판단하면 도움이 될까요?
- 엔티티의 생성자 메서드가 유효성 검증 책임을 포함하는 것과 서비스 로직이 유효성 검증을 담당하도록 하는 것 중 현업에서 어떤 것이 선호되는지 궁금합니다. 저는 전자 방식으로 코드를 작성하였는데, 그 이유는 해당 유효성 검증 로직이 엔티티 데이터 무결성을 보장과 관련이 높다면 엔티티가 책임을 맡는 것이 더 객체지향적인 측면이 강하다고 생각해서 입니다. 하지만 반대로 이런 방식이 팀 협업이나 장기적인 유지보수를 어렵게 할 수도 있을까요? 이러한 부분을 판단할 때 어떤 기준을 고려해야 할지 궁금합니다.
- 모든 테이블에 작성일시, 작성자, 수정일시, 수정자 메타데이터를 포함한다는 요구사항에 따라 Spring Data JPA의 auditing 기능을 사용했는데요. data.sql 파일로 데이터를 주입했을 때 해당 메타데이터 값을 직접 지정하지 않으면 자동으로 입력이되지 않는 현상이 있었습니다. 이 문제를 해결하기 위해 아래 방법을 시도했는데 해결하지 못했습니다.
    - adapter 모듈의 JpaConfig 클래스에 `@EnableJpaAuditing` 및 `@Configuration` 추가
    - application-dev.yml 파일에 `defer-datasource-initialization: true` 추가
   결국은 data.sql에 해당 메타데이터 값을 직접 입력했는데요. auditing 기능이 data.sql로 데이터를 초기화하는 방식에는 작동하지 않는 걸까요?

zhdiddl committed:
chore: 멀티 모듈 구성 및 빌드 관련 파일 추가

- 프로젝트 생성 후 각 모듈의 `build.gradle` 작성
- 멀티 모듈 셋팅을 위해 `settings.gradle` 추가

zhdiddl committed:
chore: 애플리케이션 및 환경 설정 파일 추가

- 애플리케이션 진입점인 `movie-api` 모듈에 `application.yml` 추가
- `movie-adapter` 모듈 아래 DB 연결 정보 로깅 레벨 설정을 포함한 설정 파일 추가
- 도커 컴포즈 사용해 DB 컨테이너를 실행할 설정 파일 추가

zhdiddl committed:
feat(domain): 엔티티 작성 및 메타데이터 필드 추출

- 기능 요구 사항에 따라 영화 조회 기능에 필요한 엔티티 추가
- 영화 관람 등급은 타입 안정성을 보장하는 enum으로 작성
- 값의 변경이 필요하지 않은 좌석 위치 정보는 VO로 작성
- 엔티티마다 포함하지만 도메인과 직접 연관이 없는 요소를 별도 추출
- 생성자, 생성일시, 수정자, 수정일시 필드를 작성하고 Jpa Auditing 적용

zhdiddl committed:
feat(application): 영화 조회 서비스 및 port 추가

- 헥사고날 아키텍처에 따라 port를 활용해 비즈니스 로직을 기술 세부 로직으로부터 독립
- `MovieServicePort` (Input Port)를 통해 외부 요청 처리
- `MovieRepositoryPort` (Output Port)를 통해 데이터를 저장소에서 조회
- `MovieService` 클래서에서 레포지토리 포트 의존성을 주입받아 비즈니스 로직 구현

zhdiddl committed:
feat(adapter): 레포지토리 생성 및 어댑터 구현

- 각 도메인의 데이터에 접근하기 위한 Jpa 레포지토리를 구현한 인터페이스 작성
- 실제 DB 접근이 필요한 영화는 `MovieJpaRepositoryAdapter`로 상세 로직 구현
- `MovieRepositoryPort`는 어댑터 덕분에 DB 변경이나 추가에 영향을 받지 않음

zhdiddl committed:
feat(adapter): 사용자 정보 자동 추가를 위해 auditing 기능 추가

- `@EnableJpaAuditing`를 활성화하여 JPA Auditing 기능을 사용하도록 설정
- 현재는 "System"으로 고정되어 있지만 실제 사용자 정보를 가져오는 로직 추가 가능

zhdiddl committed:
feat(application): 데이터 반환을 위한 dto 추가

- 클라이언트에 반환할 데이터를 위한 불변 객체 dto 추가
- 엔티티를 dto로 변환하는 팩토리 메서드를 포함
- 영화 조회시 상영 시간표를 함께 반환하도록 작성

zhdiddl committed:
feat(common): 공통 예외 처리 클래스 추가

- `CustomException`를 생성하고 `ErrorCode`에서 사용할 에러 코드를 작성
- `ErrorResponse`로 일관된 에러 응답 형식을 제공
- `GlobalExceptionHandler`를 통해 예외 처리를 중앙에서 관리

zhdiddl committed:
feat(api): 영화 목록 조회 api 구현

- `MovieController` 클래스에 영화 목록 조회 API 구현
- 클라이언트가 영화 목록을 요청할 엔드포인트 추가
- MovieServicePort` 인터페이스를 호출하여 비즈니스 로직 처리
- 기능 요구 사항에 따라 페이징 처리하지 않음

zhdiddl committed:
feat(api): 애플리케이션 진입점인 메인 클래스 설정

- 멀티 모듈 구조에 맞게 스캔 정보를 추가하고 의존성 주입하도록 작성

zhdiddl committed:
chore: 데이터베이스 초기화용 data.sql 파일 추가

- 초기 데이터 삽입을 위한 SQL 쿼리 작성
- 영화 목록 조회 기능 테스트를 위해 영화, 상영 정보, 좌석 데이터 등을 미리 삽입

zhdiddl committed:
test: 영화 목록 조회 테스트를 위한 .http 파일 추가

- 요구 사항에 따라 영화 목록 및 상영 시간표 조회 기능을 위한 .http 파일 추가

zhdiddl committed:
chore: README 작성

- 전반적인 구성 및 테이블에 대한 간략한 설명 작성

soonhankwon commented:
안녕하세요 진아님! 1주차 리뷰어를 맡게된 권순한입니다!
여러모로 깊게 고민하신 흔적이 보여서 좋습니다.
열심히 리뷰해보겠습니다 🔥

soonhankwon commented on build.gradle:
```diff
@@ -0,0 +1,46 @@
+    apply plugin: 'io.spring.dependency-management'
+
+    dependencies {
+        implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
```
모든 하위 모듈에서 jpa, web dependency를 사용하고 있는데 모듈을 나눈만큼 필요한 곳에만 의존성을 주는 것이 좋아보입니다 :)

zhdiddl replied:
공통 의존성은 롬복과 부트 스타터만 사용하고 각 모듈에서 의존성을 관리하도록 수정했습니다!

soonhankwon commented on docker-compose.yml:
```diff
@@ -0,0 +1,18 @@
+      MYSQL_ROOT_PASSWORD: root_password   # MySQL root 계정 비밀번호
+      MYSQL_DATABASE: dev_database         # 생성할 데이터베이스 이름
+      MYSQL_USER: dev_user                 # 사용자 계정 이름
+      MYSQL_PASSWORD: dev_password         # 사용자 계정 비밀번호
```
Docker-Compose 파일도 잘 작성해주셨네요.

- 다만 컨테이너도 내부에 TimeZone이 있는데(컨테이너 내부의 쉘에 들어가서 시간 확인 가능) 보통 UTC입니다.
- "운영환경"에서 DB에 데이터를 넣는데 시간이 이상하게 들어가고 조회되기도 하니 알아두시면 좋을듯 합니다.

zhdiddl replied:
말씀 주신대로 컨테이너에서 date 명령어로 확인한 결과 타임존이 UTC 였습니다. 
docker-compose의 environment 항목에 TZ: Asia/Seoul을 통해 컨테이너 시스템의 시간대를 변경해 주었습니다.
컨테이너 타임존에 대해서는 잘 몰랐는데 공부할 수 있는 기회가 되었습니다 :)

soonhankwon commented on movie-adapter/build.gradle:
```diff
@@ -0,0 +1,10 @@
+    implementation project(':movie-application') // port 사용 위해 추가
+    implementation project(':movie-domain')
+    implementation project(':movie-common')
+    implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
```
루트 모듈의 subprojects 에도 같은 의존성 설정이 있습니다. 적합한 곳에만 설정하면 좋을것 같아요 :)
- 빌드 속도 감소
- 프로젝트 구조와 의존성 관계를 이해하기 어려운 단점이 있습니다.

zhdiddl replied:
각 모듈에서 필요한 의존성 관리하도록 수정 완료했습니다!

soonhankwon commented on movie-adapter/src/main/java/com/example/movieadapter/db/MovieJpaRepository.java:
```diff
@@ -0,0 +1,9 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
spring data jpa를 사용하고 있으므로 애노테이션은 제거해주셔도 좋다고 생각합니다 :)

zhdiddl replied:
이런 실수를.. 😅 수정했습니다!

soonhankwon commented on movie-adapter/src/main/java/com/example/movieadapter/db/MovieJpaRepositoryAdapter.java:
```diff
@@ -0,0 +1,27 @@
+ */
+
+@RequiredArgsConstructor
+@Repository
```
이 부분에서 레포지토리 애노테이션을 사용해주신 것은 잘 하셨습니다 👍 
- 인터페이스를 통해서 JPA 의존성을 DIP를 통해 분리해주신 점도 좋네요!

soonhankwon commented on movie-adapter/src/main/java/com/example/movieadapter/db/MovieJpaRepositoryAdapter.java:
```diff
@@ -0,0 +1,27 @@
+    private final MovieJpaRepository movieJpaRepository;
+
+    @Override
+    public List<Movie> findAllMovies() {
```
Movie만 다루는 클래스이고 반환타입도 List<Movie>이므로 메서드 시그네처에 Movies까지 들어가는건 불필요하다고 생각하는편 입니다 :)

zhdiddl replied:
신경쓰지 못한 부분인데 정말 그렇네요! 네이밍 시 더 주의깊게 작성해야겠습니다 👍

soonhankwon commented on movie-adapter/src/main/resources/application-prod.yml:
```diff
@@ -0,0 +1,6 @@
+spring:
+  datasource:
```
charset, collate, timezone에 대한 설정이 없어서 아쉽습니다. DB를 생성할때 있어야할 기본적인 부분이라고 생각합니다 :)

zhdiddl replied:
해당 부분 반영해서 application 파일 수정했습니다!

soonhankwon commented on movie-api/src/main/java/com/example/movieapi/controller/MovieController.java:
```diff
@@ -0,0 +1,25 @@
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/movies")
```
Restful 하게 잘 url을 만들어주셨네요, 하지만 많은 곳에서 컨벤션으로 백엔드 API는 api/v1/movies 패턴을 사용합니다.

- 특히 v1(버전관리)는 현업에서 생각보다 중요하다고 경험한 점이었습니다. 이유를 한 번 생각해주셔도 좋을듯합니다.

zhdiddl replied:
url에 버전 관리 추가했습니다! 아래와 같은 장점이 있을 것이라고 예상해 보았습니다.
- 클라이언트에게 버전 별로 사용 가능한 api를 가이드하는 역할
- 유지보수 진행 과정에서 새 기능을 개발하고 테스트하는 과정에서 충돌 방지

soonhankwon commented on movie-common/src/main/java/com/example/moviecommon/exception/ErrorCode.java:
```diff
@@ -0,0 +1,32 @@
+    private final String message;
+    private final int statusCode;
+
+    ErrorCode(String message, int statusCode) {
```
Lombok을 프로젝트 전반에 사용하고 있는데 이런 보일러플레이트들은 RequiredArgsConstructor 애노테이션을 사용해도 좋지 않을까요?

zhdiddl replied:
놓친 부분이네요 😓 수정했습니다!

soonhankwon commented on movie-domain/src/main/java/com/example/moviedomain/entity/Movie.java:
```diff
@@ -0,0 +1,101 @@
+    @Column(nullable = false)
+    private LocalDate releaseDate;
+
+    @Setter private String thumbnailUrl;
```
특별하게 특정 멤버변수만 Setter를 오픈한 이유가 있을까요?
Setter를 사용해서 외부에서 "상태변경"할 가능성이 열려있습니다.
- 기본적으로 객체 내부에서 상태를 변경하도록 구현하는 것이 코드 재사용성, 캡슐화에 좋다고 생각합니다.

zhdiddl replied:
이전에 Setter 대신 상태 변경을 하는 메서드를 직접 작성했지만, 좋지 않은 리뷰를 받은 경험이 있었는데요. 😅
메서드 내부에 예외를 던지는 검증 로직을 포함했는데, 검증 책임이 해당 위치에 있는 것이 올바르지 않다는 피드백을 받았습니다. 하지만 검증 로직을 포함하지 않으면, Setter 애노테이션을 사용하는 것과 큰 차이가 없다는 생각이 들더라고요. 

코치님께서 말씀하신 “객체 내부에서 상태를 변경하도록 구현”하는 방식은 업데이트 메서드 내에 검증 로직을 포함하는 걸까요? 
```
 public void changeTitle(String newTitle) { // 제목 필드 변경 메서드
        if (newTitle == null || newTitle.isBlank()) {
            throw new IllegalArgumentException("Title cannot be empty"); // 새 제목이 null 이면 예외 처리
        }
        this.title = newTitle;
    }
```
엔티티가 검증을 하는 게 OOP적으로는 맞을 것 같다는 생각이 드는데 이 방식이 맞을지요? 이 경우에는 early-exception을 고려하지 않아도 되는지도 궁금합니다.

soonhankwon replied:
메서드 내부에 예외를 던지는 검증 로직이 있는것이 올바르지 않은 것은 다소 틀린 부분이 있는 것 같습니다. 상황에 따라 다를 수 있습니다. 책임을 누구에게 줄까? 라는 영역입니다.
- 엔티티가 검증을 하는 방법: 내 상태필드에 대한 검증은 나의 책임!
- 밸리데이터가 검증을 하는 방법: 검증에 대한 책임은 밸리데이터가 한다!
- 밸리데이터를 사용하지 않는다면, 엔티티가 검증하는 것이 올바르겠죠 :)

물론 Input값에 대한 밸리데이션의 대한 책임은 밸리데이터에게 주는것을 저는 선호합니다.(논리적으로 Validation Layer가 분리(경우의 수 감소), Early-Ex를 할 수 있음)
- 하지만, 이렇게 해도 내부에서 다양한 경우가 나오는 경우에 대한 밸리데이션도 해줘야겠죠
- 해당 경우의 수가 발생한다면, 해당 경우를 엔티티 내부에서도 검증을 하는 것이 좋다고 생각합니다.

그리고 객체 내부에서 상태를 변경하도록 구현하는 것은 검증로직 뿐이 아닙니다 :)
아래에 Setter방식과 객체 내부에서 상태변경하는 간단한 예를 써보았습니다.
- Setter를 사용하면 근본적으로 서비스 레이어에서 주요 구현이 이루어질 확률이 높습니다. -> 절차지향적
- 정말 단순한 로직과 배치성 스크립트 로직은 충분히 사용해도 좋다고 생각합니다. -> 이 경우 절차지향이 더 효율적
- Setter에 의존해서 비즈니스 로직을 만드는 습관을 가진다면, 어느순간 비대해진 서비스 레이어의 코드를 경험할 가능성이 많을꺼에요

```java
 public void changeTitle(String newTitle) { 
			 // 이 부분은 저라면 Validator에게 책임을 넘깁니다. 
			 // 이유는 Validation Layer를 명확히 해주기 위해(Input값에 대한 밸리데이션)
        if (newTitle == null || newTitle.isBlank()) {
            throw new IllegalArgumentException("Title cannot be empty"); // 새 제목이 null 이면 예외 처리
        }
        // 그런데 이부분에서 newTitle을 변환해서 저장해야한다면?
        this.title = newTitle;
    }
```

```java
 @Service
 public class bookService {
                 public void updateTitle(final String title) {
		 Book book = new Book();
		 String convertedTitle = WeirdBookChanger.convertTitle(title);
		 // 단순 Setter 사용방식 
		 // 객체 외부에서 핵심 비즈니스 로직이 진행됨(결국 서비스 레이어에서 구현)
		 book.setTitle(WeirdBookConverter.convertTitle(convertedTitle);
      }
 }
 
 @Service
 public class bookService {
                 public void updateTitle(final String title)
		 Book book = new Book();
		 Converter converter = new WeirdBookChanger();
		 // 객체에 메세지를 줘서, 내부에서 구현을 진행하는 방식
		 // 서비스에서는 객체들이 사용할 객체들만 주입, 상세 구현은 하지 않음
		 book.changeTitle(title, converter);
      }
 }
 
  public void changeTitle(String newTitle, Converter converter) { 
        if (newTitle == null || newTitle.isBlank()) {
            throw new IllegalArgumentException("Title cannot be empty"); // 새 제목이 null 이면 예외 처리
        }
        // 그런데 이부분에서 newTitle을 변환해서 저장해야한다면?
        // 메세지를 전달할 객체 참조를 받아서 메세지를 전달해서 결과를 저장
        this.title = converter.convertTitle(newTitle);
    }
```

zhdiddl replied:
"책임"을 핵심 키워드로 두고 생각해보니 검증 로직과 상태 변경 방식에 대한 기준이 더 명료해지는 것 같습니다 :) 감사합니다!!

soonhankwon commented on movie-domain/src/main/java/com/example/moviedomain/entity/Movie.java:
```diff
@@ -0,0 +1,101 @@
+    @Setter private int runtimeMinutes;
+    @Setter @Column(nullable = false) private String genre;
+
+    @OneToMany(mappedBy = "movie", fetch = FetchType.EAGER, cascade = CascadeType.ALL)
```
EAGER로는 N+1문제가 근본적으로 해결되지 않습니다 :)
인지하고 계신지 궁금하고, 이 부분에 대한 제 의견을 남기자면
- OneToMany 가 요구사항을 구현하기 위해 "꼭 필요한가?" 궁금합니다.
- 저는 해당 애노테이션들은 득보다 실이 훨씬 많다고 생각하는 편입니다.
- 경험상 해당 애노테이션을 쓰지 않고도 모든 비즈니스 로직을 구현할 수 있습니다.(e.g 중간테이블, join etc)
- 위 애노테이션(연관관계 특히 양방향)을 사용함으로써 "N+1문제"를 세심하게 고려해줘야하는 가능성이 만들어집니다. 함수로 비유하자만 "pre-condition"이 하나 더 생기는 것입니다.
- entity graph, fetch join, batch size, querydsl 등 방법으로 해결할 수 있지만, 이러한 구조를 애초에 만들지 않는 것도 하나의 방법입니다.

zhdiddl replied:
N+1 문제를 다시 검토하고 EntityGraph를 적용해 쿼리가 한 번만 출력되는 것을 확인할 수 있었습니다 :)
중간테이블의 필요성과 역할에 대해 이해하는 것이 어려워 적용까지 하지는 못했는데요.
혹시 아래 내용에 대해 간략하게라도 답변 주시면 참고해서 더 공부해 보겠습니다!

- 중간 테이블이 각 테이블의 외래키들을 포함하는 테이블을 직접 만드는 걸까요?
- “OneToMany 같은 관계 설정을 하지 않고 중간 테이블로 비즈니스 로직을 구현할 수 있다”고 리뷰 주셨는데, Jpa 가 제공하는 JoinTable 애노테이션을 활용해서 중간테이블을 만들 수 있지만 관계 설정을 꼭 포함해야 하더라고요. 말씀 주신 방법은 중간 테이블을 수동으로 직접 생성하고 외래키도 필요시 직접 업데이트하며 관리하는 방식일까요?
- OneToMany 같은 애노테이션을 사용해 외래키와 cascade 설정이 자동화되는 것보다, 외래키를 사용하지 않아서 얻는 성능 최적화의 이득이 더 크다고 보는 관점일까요?
- 중간 테이블 생성은 다대다 관계에서 주로 권장되는 것으로 알고 있었는데, 생성 기준을 어떻게 잡는 것이 올바른가요?
- “엔티티 간 연관 관계 설정에서 Pre-Condition이 생기면서 따라올 수 있는 문제”에 대해서 경험이 부족해서인지 깊게 이해되지 않은 것 같습니다. 어떤 내용을 추가로 공부하면 더 구체적으로 이해할 수 있을지 궁금합니다.

soonhankwon replied:
중간 테이블이 각 테이블의 외래키들을 포함하는 테이블을 직접 만드는 걸까요?
- 맞습니다 :)

OneToMany 같은 애노테이션을 사용해 외래키와 cascade 설정이 자동화되는 것보다, 외래키를 사용하지 않아서 얻는 성능 최적화의 이득이 더 크다고 보는 관점일까요? + Pre-Condition이 생기면서 따라올 수 있는 문제?
- cascade 설정의 자동화나 객체 그래프를 타고 사용하는 부분은 성능면에서는 별 차이가 없습니다.
- 개발에서의 "생산성"은 저에게는 아주 조금 해당 부분에서의 편리성을 느꼈습니다.
- 하지만, cascade를 DB에 직접설정하는 방법도 있구요. 
- 저는 모든 프로그램 자체를 함수로 생각하는편입니다. 예를들면
  - Pre-condition: 변수(OneToMany), 변수(ManyToOne).... -> f(나의프로그램)
  - Post-condition: f(나의프로그램) -> 결과
  - f(나의프로그램)에 변수가 들어가는한 이 변수에 대한 "각종 가능성"을 "고려"를 무조건 해줘야 합니다 결국 경우의 수가 많아진다는 것이죠.(알고리즘을 생각하시면 조금더 이해하기 편할듯 싶습니다.) 
  - 그렇다면 판단 기준 OneToMany 쓸때 생산성(굳이?) vs 각종 가능성을 고려해줘야 하는 것(혹시 잘못쓰면?)
  - 이 Trade-off에 맞춰서 판단하는것이 중요합니다. 도움이 되면 좋겠습니다 :)
 
중간 테이블 생성 및 은 다대다 관계 주로 권장?
- 맞습니다. N:N 관계에서 설정해주시면 됩니다. 보통 N:N관계에서 중간테이블이 없는 경우 "함수 종속성"이 생깁니다.
- 관계설정은 ManyToOne만 사용하는 방법도 있고, 외래키 간접참조로 구현하는 방법도 있습니다 :)
- 중간 테이블을 사용함으로써 중간테이블에 부가적인 필드를 넣으면 보다 유연하게 요구사항을 충족시킬 수 있습니다.
- eg) OrderItem 테이블에 order_id, item_id, quantity(수량)

zhdiddl replied:
지금까지는 JPA가 주는 기능을 잘 사용해야 한다는 생각에 포커스가 맞춰져 있었는데, 더 넓은 관점으로 사고할 필요성을 느꼈습니다.
앞으로도 이 부분은 지속적으로 고민하면서 공부하겠습니다. 감사합니다! 👍

soonhankwon commented on movie-domain/src/main/java/com/example/moviedomain/entity/Movie.java:
```diff
@@ -0,0 +1,101 @@
+
+    protected Movie() {}
+
+    private Movie(String title,
```
생성자를 private, 객체생성은 팩터리 메서드를 통해서만 제공하는 방법을 사용하셨네요. OOP적으로 좋은방법입니다 👍 
- 다만, WAS 을 가정하자면 밸리데이션은 최대한 "Early-Exception"을 해주는 것이 좋습니다. 들어오는 값들에 대한 검증은 "Validation Layer"에서 하도록 논리적으로 분리하는것이 좋다고 생각합니다.
- 최대한 빠르게 예외를 내주는것이 성능상에도 좋구요 :)
- 리뷰포인트, 질문에 해당 질문이 있어서 자세한 내용은 그때 보충드리도록 하겠습니다.

zhdiddl replied:
자세한 설명 감사합니다! 개념 이해에 도움이 많이 되었습니다 👍
생성자 내에 유효성 검증 로직을 제거하고 DTO에서 Spring Validation을 활용하도록 수정했습니다.

soonhankwon commented on movie-domain/src/main/java/com/example/moviedomain/valueObject/ContentRating.java:
```diff
@@ -0,0 +1,21 @@
+@Getter
+public enum ContentRating {
+
+    ALL("전체 관람가", 0),
```
ENUM으로 접근하신 것은 좋은 선택이라고 생각합니다. 👍

한 가지 소개해드릴 방법은 Converter를 사용해서 데이터베이스에 'ALL'이면 -> 'A' 이렇게 저장해서 데이터베이스 리소스를 절약하는 방법도 있습니다.

zhdiddl replied:
리팩토링하면서 적용해 보았습니다. 감사합니다 👍

soonhankwon commented:
내용이 많아서 쫌 넓은 시선으로 리뷰했습니다. PR 사이즈가 작으면 좀 더 깊게 리뷰할텐데, 아쉽습니다. 부족한 부분 생각나면 보충해드리겠습니다 :)

 ### 좋았던 점
    
  - 아키텍처에 대한 트레이드 오프를 잘 이해하고 설계한 이유가 있어서 좋았습니다.
  - 코드에 Validation 및 논리적인 과정이 보여서 너무 좋습니다.
  - 불필요한 애노테이션이 크게 없고, 잘 이해해서 사용하는 것 같습니다.
  - 처음인 멀티모듈 디자인을 무난하게 잘 구현하셨습니다.
    
  ### 아쉬웠던 점
    
 - 서비스 요구사항에 비해 과도한 아키텍처 설계라고 생각이 듭니다. 하지만 “트레이드-오프” 경험을 위해선 괜찮습니다.
  - 스프링부트, Lombok을 사용하는만큼 필요한 부분에 애노테이션 및 기능을 적절하게 사용해서 보일러 플레이트 코드를 없애서 코드를 간결하게 유지하면 더 좋을것 같습니다.
  - DDL이 없는데 추가하면 데이터베이스 스키마를 파악하는데 도움이 될 것 같습니다.
    
 ### 리뷰 포인트
    
 #### 헥사고날 아키텍처를 적용한 현재 모듈 및 패키지 구조가 올바른지 궁금합니다.
     
- 아키텍처에 대해서 공부도 많이 하신듯 하고, 아키텍처의 트레이드 오프를 고려하신 점도 인상깊습니다.
- 헥사고날 아키텍처를 이해도 명확히 잘 하신듯하고, 잘 구현하셨습니다. 하지만 “의존성 방향”에 대해서 잘 생각하시고 개선되면 좋을 듯 합니다.
- 다만, 고민하신대로 구조가 세분화된 만큼 초반부터 어느 시점까지 경험상 생산성이 “매우” 떨어지고 관리할 포인트도 많아집니다.
- 아키텍처의 본질은 결국 “제약”이기 때문입니다. 어떠한 모듈은 “독립적”이고 “변화에 유연”해야 한다고 생각되면 이 모듈은 최대한 의존성이 없도록 “제약”을 걸어주는 거죠.
- 하지만 아키텍처 때문에 정말 중요한 본질인 “비즈니스 요구사항, 생산성”을 저하된다면 심플한 아키텍처로 진행하는 것이 더 좋습니다.
- 팀 내부에서 아키텍처에 대한 의견 통일 및 “컨벤션”을 확립하지 않으면 문제가 생길 여지도 많습니다.
- 다음 주차에서 해당 아키텍처를 유지 또는 변경해보면서 어떠한 “트레이트-오프”가 있는지 경험하시면 좋을 것 같습니다.

### 기타 질문

- **UseCase 적용**
    - 복잡한 비즈니스 규칙이 생길 때 UseCase를 도입하면 유리합니다. 이와 관련해 **Facade 디자인 패턴**을 참고해보시면, 비즈니스 로직의 캡슐화와 단순화에 도움이 될 수 있습니다.
    - UseCase는 반드시 초기에 적용해야 할 필요가 없습니다. 현재 요구사항이 단순하다면, 필요할 때 적시에 도입하면 됩니다.
    - 현재 비즈니스 요구가 심플하기 때문에 **적용하지 않아도 충분히 무방**하다고 생각합니다.
- **`api` 모듈에서 `adapter` 직접 의존**
    - Bean 스캔 문제인지, 의존성이 없어서 생기는 에러 로그인지 파악하시면 도움이 될 듯합니다.
    - 문제가 Bean 스캔문제라면 componentscan으로 해결이 될 것 같은데, 안된다고 하셨으니 의존성 문제인 것 같습니다.
        - 모듈이 여러 개인 만큼 저도 코드를 자세히 들여다봐야 문제를 알 수 있을것같습니다
    - 하지만, 저라면은 비즈니스 요구가 심플하니, 현재구조에서 adapter를 없애고 infastructure 모듈로 통합해서 진행할 듯 합니다.
    - 그렇다면 infastructure에 JPA 및 레포지토리 구현체 , domain모듈의 port에 레포지토리 인터페이스를 선언해서 api 모듈이 domain 모듈만 의존하도록 구조자체를 심플하게 바꿔서 설계할 것 같습니다.
- **`adapter` 모듈에서 `application` 직접 의존**
    - 위에서 답변드린  부분과 비슷하네요.
    - 그런데 adapter가 application을 의존하고 있다면 의존성 방향을 한 번 다시 고민해보는게 좋을듯하네요.


- **`application.yml` 파일 위치**
    - 크게 문제가 없습니다. 모듈을 독립적으로 gradle build하는 경우도 있을 수 있기 때문에요.
    - 하지만, 한 군데서 통합적으로 관리하는게 관리면에서는 용이하다고 생각합니다.
    - profile include 같은 디테일한 설정을 생각해볼 수 있습니다 🙂


- 엔티티에서 `@Column(nullable = false)`를 설정할 때 어떤 기준으로 추가 여부를 판단하는면 좋을까요?
    - nullable false 같은 속성은 실제 데이터베이스에 들어가는 속성들은 코드에 명시해 주면 줄수록 개발에 용이하다고 생각합니다.
    - 생성자를 보는것과, 엔티티 Column을 보는 것 둘 중에 어느것이 더 일반적으로 의도가 명확하게 전달될까 생각해보시면 좋을 것 같습니다.
    - 예를들면 말씀하신 영화 제목은 String 타입인데, 실제로 해당 필드를 null 로 생성하는 코드를 어디선가 넣어서 구현해도 아무런 컴파일 에러가 나지 않습니다. 이 부분은 런타임에서 예외가 나서 이미 Stage 또는 Production에 나가서야 모니터링 될 것입니다.
    - 제가 좋아하는 방법은 고전적이지만 “assert”를 코드 중간 중간 넣어서 저의 의도를 표시해주는 것입니다. sentinal이라고 하는 방법인데 한번 알아보시면 좋을 듯합니다!
- 엔티티 생성자를 of 같은 메서드명을 가진 팩토리 메서드로 제공하는 것에 대한 트레이드 오프를 어떻게 생각하시는지 궁금합니다.
    - 좋은 방법입니다. 메서드 레퍼런스도 편하게 사용할 수 있고, 메서드 네이밍도 의도를 전달할 수 있고, 함수형 스타일을 사용하기에도 편리합니다.
    - 하지만, 팀에서 “컨벤션”으로 정확하게 네이밍 규칙을 정해놓지 않으면 중구난방이 될 수 있어서 이것을 꼭 정해놓는 것이 중요합니다.
    - 그리고 엔티티 또는 도메인에 비즈니스 로직보다 비대한 객체를 생성하는 코드가 많아질 수 있습니다. 이런 부분들로 인해 빌더패턴을 조합해서 사용해도 좋습니다.
    - 결국 팀에서의 컨벤션 차이라고 생각합니다.
- 엔티티의 생성자 메서드가 유효성 검증 책임을 포함하는 것과 서비스 로직이 유효성 검증을 담당하도록 하는 것 중 현업에서 어떤 것이 선호되는지 궁금합니다.
    - 팀마다 다르다고 생각합니다.
    - 저는 “Validation Layer”를 Service Layer와 논리적으로 구분하는 것을 중요하게 여기고 선호합니다. 따라서 생성자에 해당 책임을 주지 않습니다.
    - 저는 보통 Service Layer에서는 밸리데이션을 하지 않습니다. 여기에 들어오는 Input은 적합한 Value(eg not null)이며, assert로 not null임을 표시해주는 식으로 의도를 전달합니다.
    - Validation Layer에서는 Input값에 대한 모든 검증을 진행해서 함수로 따지면 제가 생각한 pre-condition을 만들어줍니다. 여기서 버그가 나오는 것은 계속 추가가 되겠지요.
    - 또한 스프링의 Validation 을 활용할 수 있기 때문에 Valdation Layer에서 훨씬 빠르게 Early-Exception을 할 수 있어, 성능적으로도 좋습니다.
    - 그리고 “논리적”으로도 명확한 프로그래밍을 할 수 있고 문제가 생길경우 디버깅 포인트 파악도 용이해집니다.
    - 그리고 불필요한 NPE 체크도 안하게 되구요. NPE 체크 같은 경우는 논리적으로 명확하지 않다면, 코드에 NPE 방어 코드가 난립하게 되는 문제가 생깁니다.

zhdiddl committed:
refactor: 각 모듈의 불필요한 의존성 제거

- 모든 하위 모듈에 공통 적용한 의존성 중 불필요한 내용 제거
- 각 하위 모듈에 필요한 의존성을 개별적으로 추가

zhdiddl committed:
refactor: MySQL 컨테이너 시간대 및 문자 관련 설정

- 시간대 설정을 추가해 운영환경에서 DB에 데이터를 변경 및 조회하는 경우 혼란 방지
- 한글 인코딩에 문제 없도록 문자열 인코딩 및 정렬 규칙을 추가

zhdiddl committed:
chore: 불필요한 주석 및 애노테이션 제거

- JpaRepository를 상속하는 인터페이스에 불필요한 애노테이션 제거
- 작성한 코드 기능에 적절하지 않다고 판단되는 주석 삭제
- 빈 스캔 관련 중복으로 설정된 코드 삭제

zhdiddl committed:
refactor: 롬복을 활용해 보일러플레이트 코드 제거

- 롬복 애노테이션을 통해 private final 필드를 초기화하는 생성자를 자동 생성

zhdiddl committed:
refactor: 간결한 메서드명으로 수정

- 메서드 반환타입이 이미 List<Movies>로 명시된 점을 반영
- 직관적이고 간결한 메서드명을 사용하기 위해 `findAll()`로 수정
- 해당 메서드를 사용한 코드도 함꼐 수정

zhdiddl committed:
refactor: 버전 관리를 위한 url 컨벤션 추가

- api/v1 패턴을 사용해 api의 버전 관리
- 클라이언트에게 버전 별로 사용 가능한 api를 가이드
- 유지보수 과정에서 새 기능 개발 및 테스트 시 오류 및 충돌 방지

zhdiddl committed:
feat(common): enum 값을 축약해 db에 저장하는 컨버터 추가

- `ContentRating`enum 값을 db에 저장시 컨버터로 리소스 절약
- Jpa가 제공하는 Converter 애노테이션을 활용
- 저장 시 축약된 문자열로 변환하고 조회 시 다시 enum으로 변환
- 커스텀 익셉션을 추가하고 변환 로직에 유효성 검증에 반영
- 데이터 초기화 파일에도 축약값으로 변환할 수 있는 쿼리 추가

zhdiddl committed:
fix: 컨트롤러 및 테스트 파일 상 url 수정

- api 경로 수정 및 누락된 '/' 추가

zhdiddl committed:
refactor: collation 설정 업데이트

- MySQL 8.0 버전부터 제공되는 `utf8mb4_0900_ai_ci`으로 대체

zhdiddl committed:
refactor: 모듈 구조 재정비 및 코드 리팩토링

- 기존 모듈 구조에서 불필요한 모듈 병합 및 제거 처리
- 헥사고날 아키텍처를 기반으로 하되 생산성을 저해하지 않도록 재정비
- `adapter`를 `infrastructure`에 포함시켜 구조 단순화
- 구조 리팩토링 과정에서 대량의 파일 이동, 추가 및 삭제 발생
- 이에 따른 package 및 import 문 수정
- 변경된 모듈 구조 변경을 반영해 README 파일 수정

zhdiddl committed:
chore: API URL에 v1 버전 추가

- 기존 API 엔드포인트에 버전 정보(v1) 추가
- 향후 버전 관리를 고려하여 명확한 URL 구조 적용

zhdiddl committed:
refactor: 엔티티 생성 메서드에서 검증 로직 제거

- 엔티티 생성 과정에서 수행하던 유효성 검사를 제거
- Validation Layer에서 검증을 수행하도록 추후 변경 예정

zhdiddl committed:
perf: N+1 문제 해결을 위해 @EntityGraph 적용 및 DTO 최적화

- Movie 조회 시 연관 데이터를 함께 로딩하도록 @EntityGraph 추가
- 데이터 조회 성능 최적화 및 쿼리 수 감소
- 필요한 데이터를 선택적으로 반환하기 위해 DTO 최적화

zhdiddl committed:
feat: 영화 목록 정렬 조회 및 생성 기능 추가

- findAll 메서드에 Sort 파라미터 추가하여 정렬 기능 지원
- 영화 목록을 개봉일 내림차순으로 정렬하여 조회하는 메서드 구현
- MovieRepositoryPort를 사용하여 영화 저장 및 조회 수행
- 새 영화 저장 시 DTO에 적용된 Spring Validation으로 유효성 검증

zhdiddl committed:
refactor: ContentRatingConverter에 assert 추가 및 기본값 적용

- 불필요한 NPE를 줄이기 위해 코드 개선
- ContentRating이나 DB 저장 값이 null이 아닌지 assert 검증 추가
- 불러온 DB 저장 값이 null일 경우 기본 값 적용

zhdiddl committed:
feat: DDL 파일 추가

- 초기 스키마를 정의하기 위한 DDL 파일 작성
- DB 자원 절약을 위해 BIGINT 대신 USIGNED INT 사용
- 성능 최적화를 고려해 외래키 제약 조건 제거

zhdiddl reviewed (commented):
정성스러운 리뷰 감사합니다! 말씀해 주신 부분 고민하면서 계속 공부해보겠습니다 🚀

zhdiddl committed:
chore: DB 설정 개선 및 추가

- 환경 별로 charset, collation, timezone 설정 추가
